### PR TITLE
feat: add title action for announce sign

### PR DESCRIPTION
**Pull request by layertwo**
### Overview
The Announce sign currently only announce to a player's chat. Minecraft also has the `sendTitle` feature, where a title can be display on the player's screen. To not ruin backwards compatibility, I've created a method for determine if the announce is default, or if it is to use the title method.

A user can provide this option by adding an additional string "title" to the second line of a sign. Example
```
[train]
announce title
foo
bar
```

#### Strings are separate messages
Each message line on a sign is now sent as a separate message to players, rather than concatenating the strings together and sending that as a single message.

The current method concatenates all strings and sends them to the chat as a single message. This ends up being an issue because separate lines are merged into a single line and make it hard to read. Example: using "foo" on line 3 and "bar" on line 4 ends up sending to the player chat "foobar" rather than "foo bar".

#### AnnounceMessage
Created a private `AnnounceMessage` class that handles both title and subtitle (need multiple messages for `sendTitle`), and can be applied for sending a message to the player's chat as well.

This class provides a clean method for both sending to chat and sending title and subtitle to a player, and was the best path forward I considered for deconflicting having to spend cycles generating the message per player (instead it can be calculated ahead of time and only recalled when it needs to be sent).

### CODE REVIEW

**Comment by bbayu123** [comment]
Two issues that I can see here
1. This change fundamentally breaks all existing announce signs that have more than 2 lines of messages, thus you are throwing out backwards compatibility.
2. The title and subtitle messages completely bypass the alias and variable handling code that TrainCarts has, so aliases and variables won't get translated when using titles and subtitles. 

If the logic is sufficiently different from the existing announce sign, it's better to create a new sign rather than adding functionality to an existing sign.

**Comment by layertwo** [comment]
Good call outs!

I definitely went back and forth when designing this trying to figure out how to keep the existing behavior and also be ablee to use `sendTitle`. The major reason for the `AnnounceMessage` class is that I didn't want to calculate the message during send time, and only try to calculate it once outside of the sending logic.

For 1: I think this could be solvable with an additional property in `AnnounceMessage`, and then send each new line as a message. The major change is here is that each line is sent separately and not as a single message. This is probably the larger discussion to be had.

For 2: This is totally fair. All depends on what the intended behavior of this sign is and if this change is acceptable.

Thanks for taking a look, more than happy to iterate on this and provide something that we all agree on.

File: src/main/java/com/bergerkiller/bukkit/tc/signactions/SignActionAnnounce.java [hunk 1]
```diff
@@ -37,16 +37,21 @@ public boolean match(SignActionEvent info) {
...
+            SignActionAnnounce.useTitle = isTitleAnnounce(info);
+
+            if (info.isTrainSign() && info.isAction(SignActionType.GROUP_ENTER)) {
+                if (!info.hasRailedMember() || !info.isPowered()) return;
+                sendMessage(info.getGroup(), message);
+            } else if (info.isCartSign() && info.isAction(SignActionType.MEMBER_ENTER)) {
+                if (!info.hasRailedMember() || !info.isPowered()) return;
+                sendMessage(info.getMember(), message);
+            } else if (info.isRCSign()) {
+                info.getRCTrainGroups().forEach(group -> sendMessage(group, message));
```
Comment (bbayu123) [inline src/main/java/com/bergerkiller/bukkit/tc/signactions/SignActionAnnounce.java:54]: @bergerkiller Is this slower than a plain for-loop? Will it create a significant slowdown if this is done?

File: src/main/java/com/bergerkiller/bukkit/tc/signactions/SignActionAnnounce.java [hunk 2]
```diff
@@ -66,9 +71,29 @@ public boolean build(SignChangeActionEvent event) {
                 .setPermission(Permission.BUILD_ANNOUNCER)
                 .setName("announcer")
                 .setDescription(event.isRCSign() ?
-                        "remotely send a message to all the players in the train" :
-                        "send a message to players in a train")
+                        "announce a message to all the players in the train" :
```
Comment (bbayu123) [inline src/main/java/com/bergerkiller/bukkit/tc/signactions/SignActionAnnounce.java:74]: Right now it's not obvious from the sign placement message whether an RC sign or a normal sign is placed. This is not ideal.

**Comment by bbayu123** [comment]
One last thing that I can think of:
Since you currently don't allow for specifying a fade-in, stay, and fade-out time, so someone will eventually ask for it. What are the possible ways of extending this - that you can think of - so that the 3 values can be provided?

**Comment by layertwo** [comment]
Hmmm... good question. I think there's enough space to add this to line 2 with something like 20/70/10, but that'll probably end up feeling cramped.

File: src/main/java/com/bergerkiller/bukkit/tc/signactions/SignActionAnnounce.java (hunk 1 above)
Comment (bergerkiller) [reply src/main/java/com/bergerkiller/bukkit/tc/signactions/SignActionAnnounce.java:54]: Nah that's fine, Collection.forEach isn't slower than a for loop, that's only an issue with stream api but even here that wouldn't have been a problem

**Comment by bergerkiller** [comment]
just a thought, why isn't it possible to add a new sign action completely called "showtitle" which does all the title handling? I kind of also want this because of build permission rules and stuff. But your solution is fine too.

Don't know about the fade-in time and stuff, that's difficult to specify. Would've been easier if it wasn't constrained to what can fit on a sign. But if the unit is in ticks, I think its fine to allow something like "showtitle 5 40 5" and stuff. That's not too long

**Comment by bbayu123** [comment]
That's what I said as well
> If the logic is sufficiently different from the existing announce sign, it's better to create a new sign rather than adding functionality to an existing sign.

**Comment by layertwo** [comment]
I think the consensus here is that a separate title sign is the best path forward. That works for me.

@NobodyBestFriend can you post the sign syntax like how I have in the overview of this PR? I'm not sure I exactly follow what you're thinking.

I'll close this PR and open a new one for a title sign.

**Pull request closed without merging**